### ISSUE TITLE
Unable to create Hamiltonian from geometry

### ISSUE THREAD

**Issue by juijan** [zerothi/sisl#603]
Hey Nick,

I exprience the error below when trying to create a Hamiltonian from a geometry. Any idea why this might happen?

I re-install sisl from conda-forge the error ocurred in 0.13.0 and 0.12.0.

**Version info**
```
$ conda install sisl==0.13.0

$ python -c 'import sys; print(sys.version)'
3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0]

$ python -c 'import sisl; print(sisl.__version__)'
0.0.0.dev+851d78c55d40d1f7aa1f5d8320aae1d3364064b8

```

**The script:**
```
import sisl
H = sisl.Hamiltonian(sisl.geom.graphene())
```

**The error:**
```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[10], line 1
----> 1 H = sisl.Hamiltonian(sisl.geom.graphene())

File [~/mambaforge/envs/jupyter/lib/python3.10/site-packages/sisl/physics/hamiltonian.py:52](https://file+.vscode-resource.vscode-cdn.net/home/ICN2/nwittemeier/Projects/2023/tb-bdg-model/~/mambaforge/envs/jupyter/lib/python3.10/site-packages/sisl/physics/hamiltonian.py:52), in Hamiltonian.__init__(self, geometry, dim, dtype, nnzpr, **kwargs)
     50 def __init__(self, geometry, dim=1, dtype=None, nnzpr=None, **kwargs):
     51     """ Initialize Hamiltonian """
---> 52     super().__init__(geometry, dim, dtype, nnzpr, **kwargs)
     53     self._reset()

File [~/mambaforge/envs/jupyter/lib/python3.10/site-packages/sisl/physics/sparse.py:617](https://file+.vscode-resource.vscode-cdn.net/home/ICN2/nwittemeier/Projects/2023/tb-bdg-model/~/mambaforge/envs/jupyter/lib/python3.10/site-packages/sisl/physics/sparse.py:617), in SparseOrbitalBZSpin.__init__(self, geometry, dim, dtype, nnzpr, **kwargs)
    614     spin = kwargs.pop('spin')
    615 self._spin = Spin(spin, dtype)
--> 617 super().__init__(geometry, len(self.spin), self.spin.dtype, nnzpr, **kwargs)
    618 self._reset()

File [~/mambaforge/envs/jupyter/lib/python3.10/site-packages/sisl/physics/sparse.py:71](https://file+.vscode-resource.vscode-cdn.net/home/ICN2/nwittemeier/Projects/2023/tb-bdg-model/~/mambaforge/envs/jupyter/lib/python3.10/site-packages/sisl/physics/sparse.py:71), in SparseOrbitalBZ.__init__(self, geometry, dim, dtype, nnzpr, **kwargs)
     68     dim = dim + 1
     70 # Initialize the sparsity pattern
---> 71 self.reset(dim, dtype, nnzpr)
     72 self._reset()

File [~/mambaforge/envs/jupyter/lib/python3.10/site-packages/sisl/sparse_geometry.py:118](https://file+.vscode-resource.vscode-cdn.net/home/ICN2/nwittemeier/Projects/2023/tb-bdg-model/~/mambaforge/envs/jupyter/lib/python3.10/site-packages/sisl/sparse_geometry.py:118), in _SparseGeometry.reset(self, dim, dtype, nnzpr)
    116 # query dimension of sparse matrix
    117 s = self._size
--> 118 self._csr = SparseCSR((s, s * self.geometry.n_s, dim), nnzpr=nnzpr, dtype=dtype)
    120 # Denote that one *must* specify all details of the elements
    121 self._def_dim = -1

File [~/mambaforge/envs/jupyter/lib/python3.10/site-packages/sisl/sparse.py:144](https://file+.vscode-resource.vscode-cdn.net/home/ICN2/nwittemeier/Projects/2023/tb-bdg-model/~/mambaforge/envs/jupyter/lib/python3.10/site-packages/sisl/sparse.py:144), in SparseCSR.__init__(self, arg1, dim, dtype, nnzpr, nnz, **kwargs)
    141 elif isinstance(arg1, (tuple, list)):
    143     if isinstance(arg1[0], Integral):
--> 144         self.__init_shape(arg1, dim=dim, dtype=dtype,
    145                           nnzpr=nnzpr, nnz=nnz,
    146                           **kwargs)
    148     elif len(arg1) != 3:
    149         raise ValueError(self.__class__.__name__ + ' sparse array *must* be created '
    150                          'with data, indices, indptr')

File [~/mambaforge/envs/jupyter/lib/python3.10/site-packages/sisl/sparse.py:236](https://file+.vscode-resource.vscode-cdn.net/home/ICN2/nwittemeier/Projects/2023/tb-bdg-model/~/mambaforge/envs/jupyter/lib/python3.10/site-packages/sisl/sparse.py:236), in SparseCSR.__init_shape(self, arg1, dim, dtype, nnzpr, nnz, **kwargs)
    231 self._nnz = 0
    233 # Important that this is zero
    234 # For instance one may set one dimension at a time
    235 # thus automatically zeroing the other dimensions.
--> 236 self._D = zeros([nnz, K], dtype)
    238 # Denote that this sparsity pattern hasn't been finalized
    239 self._finalized = False

AttributeError: 'SparseCSR' object has no attribute '_D'
```

**Comment by zerothi** [comment]
Hi, odly enough this has always been a bug in sisl, but no code coverage uncovered this until recently, it got fixed in 1a06052

It is very puzzling to me it was never something that showed up... Thanks for the report!

Please try with the dev version, if it still does not work, please reopen!

**Comment by juijan** [comment]
Yep, works fine with the dev version.

Thanks!
